Fix(#23651): Trigger library search on filter change

AmiyaPiyush1 opened the pull request:
## Overview

1. This PR fixes  #23651.

2. This PR fixes the bug where the Category and Language filters on the Community Library page did not work. This was because the dropdowns were not configured to trigger a search refresh when their values changed. 

   This PR connects the dropdowns to the existing search function. Now, changing the Category or Language will correctly filter the search results, just as the text search bar does.

3. The original bug occurred because: The functions that handle dropdown selections (`toggleSelection()` and `deselectAll()` in `search-bar.component.ts`) did not include a call to the `onSearchQueryChangeExec()` function. 

   As noted by a maintainer in the issue comments, this was not a regression (a feature that broke) but rather missing functionality from the original implementation.

## Essential Checklist

- [x] I have linked the issue that this PR fixes in the "Development" section of the sidebar.
- [x] I have checked the "Files Changed" tab and confirmed that the changes are what I want to make.
- [x] I have written tests for my code.
- [x] The **PR title** starts with "Fix #bugnum: " or "Fix part of #bugnum: ...", followed by a short, clear summary of the changes.
- [x] I have assigned the correct reviewers to this PR (or will leave a comment with the phrase "@{{reviewer_username}} PTAL" if I can't assign them directly).

## Proof that changes are correct

Here is the new demo video as requested by @NITISH084, showing the text, category, and language filters working correctly by themselves and all together.

**https://res.cloudinary.com/dsbmjh24b/video/upload/v1762344725/fep4mugo7nforciqyrid.mp4**

NITISH084 commented:
@AmiyaPiyush1  I reran the failing CI tests yesterday, but it still seems one CI test is failing. Please see it and do the same as you did for confetti (file an issue for it if it's not already filed).

AmiyaPiyush1 commented:
@NITISH084 I filed the CI failure report: #23808. PTAL,Thanks!

NITISH084 commented:
@AmiyaPiyush1  this is Stack trace, Please update this in the issue description: 
```FAIL core/tests/puppeteer-acceptance-tests/specs/logged-out-learner/visits-the-community-library-and-look-around.spec.ts (203.186 s)
  Logged-Out Learner
    ✕ should be able to discover the community library
    ✕ should be able to look at different categories in community library

  ● Logged-Out Learner › should be able to discover the community library

    TimeoutError: waiting for selector `span.oppia-unique-progress-id` failed: timeout 30000ms exceeded

      2651 |       });
      2652 |       await this.clickOnElementWithSelector(explorationConfirmPublishButton);
    > 2653 |       await this.page.waitForSelector(explorationIdElement);
           |                       ^
      2654 |       const explorationIdUrl = await this.page.$eval(
      2655 |         explorationIdElement,
      2656 |         element => (element as HTMLElement).innerText

      at new WaitTask (../../../node_modules/puppeteer/src/common/DOMWorld.ts:813:28)
      at DOMWorld.waitForSelectorInPage (../../../node_modules/puppeteer/src/common/DOMWorld.ts:656:22)
      at Object.internalHandler.waitFor (../../../node_modules/puppeteer/src/common/QueryHandler.ts:78:19)
      at DOMWorld.waitForSelector (../../../node_modules/puppeteer/src/common/DOMWorld.ts:511:25)
      at Frame.waitForSelector (../../../node_modules/puppeteer/src/common/FrameManager.ts:1290:47)
      at Page.waitForSelector (../../../node_modules/puppeteer/src/common/Page.ts:3222:29)
      at confirmPublish (utilities/user/exploration-editor.ts:2653:23)
          at runMicrotasks (<anonymous>)
      at BaseUser.publishExplorationWithMetadata (utilities/user/exploration-editor.ts:2687:14)

  ● Logged-Out Learner › should be able to look at different categories in community library

    TimeoutError: waiting for selector `span.oppia-unique-progress-id` failed: timeout 30000ms exceeded

      2651 |       });
      2652 |       await this.clickOnElementWithSelector(explorationConfirmPublishButton);
    > 2653 |       await this.page.waitForSelector(explorationIdElement);
           |                       ^
      2654 |       const explorationIdUrl = await this.page.$eval(
      2655 |         explorationIdElement,
      2656 |         element => (element as HTMLElement).innerText

      at new WaitTask (../../../node_modules/puppeteer/src/common/DOMWorld.ts:813:28)
      at DOMWorld.waitForSelectorInPage (../../../node_modules/puppeteer/src/common/DOMWorld.ts:656:22)
      at Object.internalHandler.waitFor (../../../node_modules/puppeteer/src/common/QueryHandler.ts:78:19)
      at DOMWorld.waitForSelector (../../../node_modules/puppeteer/src/common/DOMWorld.ts:511:25)
      at Frame.waitForSelector (../../../node_modules/puppeteer/src/common/FrameManager.ts:1290:47)
      at Page.waitForSelector (../../../node_modules/puppeteer/src/common/Page.ts:3222:29)
      at confirmPublish (utilities/user/exploration-editor.ts:2653:23)
          at runMicrotasks (<anonymous>)
      at BaseUser.publishExplorationWithMetadata (utilities/user/exploration-editor.ts:2687:14)

Test Suites: 1 failed, 1 total
Tests:       2 failed, 2 total
Snapshots:   0 total
Time:        204.904 s
Ran all test suites matching /core\/tests\/puppeteer-acceptance-tests\/specs\/logged-out-learner\/visits-the-community-library-and-look-around.spec.ts/i.
```

AmiyaPiyush1 commented:
Hi @Nitish084,

I've investigated the failing CI checks, and here is my analysis and plan.

The main failure is in the Full-stack tests / Acceptance (logged-out-learner/visits-the-community-library-and-look-around) job. After analyzing the logs, I found there are two distinct errors happening.

Failure 1: Flaky Test (Welcome Modal)
The log first shows a timeout error when the test tries to create its second and third explorations:

Error: [test-log]: welcome modal not found: waiting for selector \button.e2e-test-dismiss-welcome-modal` failed: timeout 10000ms exceeded`

Analysis: This is a flaky test issue. The welcome modal only appears on the first exploration. The test isn't set up to handle the modal not appearing on subsequent runs, so it waits for 10 seconds and times out.

Action: This seems unrelated to my PR's changes. I have filed a separate new issue for this (#23813) so it can be tracked and fixed independently.

Failure 2: Crash from PR Logic (Main Failure)
The second error is the one that actually crashes the test suite and is related to my PR.

Error Stack Trace:

FAIL core/tests/puppeteer-acceptance-tests/specs/logged-out-learner/visits-the-community-library-and-look-around.spec.ts (196.96 s)
  Logged-Out Learner
    ✕ should be able to look at different categories in community library (3291 ms)

  ● Logged-Out Learner › should be able to look at different categories in community library

    Execution context was destroyed, most likely because of a navigation.

      3453 |
      3454 |     for (const option of filterOptions) {
    > 3455 |       const optionText = await this.page.evaluate(
           |                          ^
      ...
      at LoggedOutUser.filterLessonsByCategories (utilities/user/logged-out-user.ts:3455:26)
      at Object.<anonymous> (specs/logged-out-learner/visits-the-community-library-and-look-around.spec.ts:70:5)
Analysis: This is caused by my fix. My PR correctly triggers a new search (a page navigation) as soon as a filter is changed. The acceptance test, however, doesn't expect this navigation. It clicks the filter and immediately tries to read text from the page, but my code has already started navigating away, which causes the Execution context was destroyed crash.

My Plan
I have already filed the issue for Failure 1.

I will now fix Failure 2 as part of this PR, since my code caused it.

To fix this, I will update the acceptance test logic (in utilities/user/logged-out-user.ts) to correctly await the page navigation after it clicks a filter, before it tries to perform its next action.

Is this plan good to you? Thanks!

jayam04 commented:
> Hi @NITISH084,
> 
> I've investigated the failing CI checks, and here is my analysis and plan.
> 
> The main failure is in the Full-stack tests / Acceptance (logged-out-learner/visits-the-community-library-and-look-around) job. After analyzing the logs, I found there are two distinct errors happening.
> 
> Failure 1: Flaky Test (Welcome Modal) The log first shows a timeout error when the test tries to create its second and third explorations:
> 
> Error: [test-log]: welcome modal not found: waiting for selector \button.e2e-test-dismiss-welcome-modal` failed: timeout 10000ms exceeded`
> 
> Analysis: This is a flaky test issue. The welcome modal only appears on the first exploration. The test isn't set up to handle the modal not appearing on subsequent runs, so it waits for 10 seconds and times out.
> 
> Action: This seems unrelated to my PR's changes. I have filed a separate new issue for this (#23813) so it can be tracked and fixed independently.
> 
> Failure 2: Crash from PR Logic (Main Failure) The second error is the one that actually crashes the test suite and is related to my PR.
> 
> Error Stack Trace:
> 
> FAIL core/tests/puppeteer-acceptance-tests/specs/logged-out-learner/visits-the-community-library-and-look-around.spec.ts (196.96 s) Logged-Out Learner ✕ should be able to look at different categories in community library (3291 ms)
> 
> ● Logged-Out Learner › should be able to look at different categories in community library
> 
> ```
> Execution context was destroyed, most likely because of a navigation.
> 
>   3453 |
>   3454 |     for (const option of filterOptions) {
> > 3455 |       const optionText = await this.page.evaluate(
>        |                          ^
>   ...
>   at LoggedOutUser.filterLessonsByCategories (utilities/user/logged-out-user.ts:3455:26)
>   at Object.<anonymous> (specs/logged-out-learner/visits-the-community-library-and-look-around.spec.ts:70:5)
> ```
> 
> Analysis: This is caused by my fix. My PR correctly triggers a new search (a page navigation) as soon as a filter is changed. The acceptance test, however, doesn't expect this navigation. It clicks the filter and immediately tries to read text from the page, but my code has already started navigating away, which causes the Execution context was destroyed crash.
> 
> My Plan I have already filed the issue for Failure 1.
> 
> I will now fix Failure 2 as part of this PR, since my code caused it.
> 
> To fix this, I will update the acceptance test logic (in utilities/user/logged-out-user.ts) to correctly await the page navigation after it clicks a filter, before it tries to perform its next action.
> 
> Is this plan good to you? Thanks!

@AmiyaPiyush1 I see two issues filed, though none is for failure 2. Does that mean there are 3 failure, out of which one is caused by your changes?

AmiyaPiyush1 committed:
fix(library): Fix filter test and trigger search on dropdown close

AmiyaPiyush1 commented:
Hi @NITISH084,

I’ve fixed the previous bug and successfully filed the related issue for it.  
I’ve also filed a new issue for the current CI failure (the ChromeHeadless disconnection flake).  

Could you please guide me on what I should do next?

Thank you!

AmiyaPiyush1 committed:
fix(library): Fix filter test and trigger search on dropdown close

AmiyaPiyush1 commented:
Acceptance (exploration-editor/load-complete-and-restart-exploration-preview)

The failing CI test was covered in issue #23828

AmiyaPiyush1 commented:
@Nik-09  PTAL

regards Amiya Piyush

Nik-09 commented on core/templates/pages/library-page/search-bar/search-bar.component.ts:
```diff
@@ -185,13 +185,22 @@ export class SearchBarComponent implements OnInit, OnDestroy {
+   * Triggers a search query when the dropdown menu is closed.
+   * @param {boolean} isOpen - The new open state of the dropdown.
+   */
+  onDropdownClose(isOpen: boolean): void {
```
Can you please rename this method? Currently, `onDropdownClose` is very generic menthod name and does not helps to understand the context and functionality.

Nik-09 commented on core/templates/pages/library-page/search-bar/search-bar.component.html:
```diff
@@ -41,80 +42,80 @@
-             *ngIf="!first && !last"
-             class="dropdown-item">
+          <div (click)="toggleSelection('categories', item.id)"
+               (keydown.enter)="toggleSelection('categories', item.id)"
```
I notice that each line has two leading spaces. Could you please remove them?

AmiyaPiyush1 replied:
Done. I have renamed the method to triggerSearchOnDropdownClose to make it more descriptive and specific to its functionality.

AmiyaPiyush1 replied:
Done. I have removed the extra leading spaces and fixed the indentation for these lines.

Nik-09 reviewed (changes requested):
Hi @AmiyaPiyush1, thank you for the code changes.
I've added two minor comments, PTAL.

Thanks

AmiyaPiyush1 committed:
Fix(library): Rename dropdown method, fix HTML indentation and update tests

AmiyaPiyush1 committed:
Fix HTML indentation and line length

AmiyaPiyush1 committed:
Fix HTML indentation and line length

AmiyaPiyush1 commented:
The following CI failures were covered in given issue numbers - 
Full-stack tests / E2E (wipeout) (pull_request) -  #23920


Full-stack tests / Lighthouse perf (shard 1) (pull_request -  #23921

AmiyaPiyush1 commented:
Hi @Nik-09,
I've addressed all the requested changes — including renaming the dropdown method, fixing the HTML indentation, and updating the related lines.

I also reintroduced the anchor tags in place of div because the issue was resolved even with them, and I wanted to keep the changes minimal while preserving the original structure.

All related CI failures have been reported in their respective issues, and the PR is now updated and ready for review.

PTAL when you get a chance — thank you!

AmiyaPiyush1 committed:
Merge branch 'develop' of https://github.com/AmiyaPiyush1/oppia into fix-library-filter-23651

AmiyaPiyush1 committed:
Resolve merge conflict: Update CSS to float-start

AmiyaPiyush1 committed:
Merge branch 'develop' into fix-library-filter-23651

Nik-09 reviewed (changes requested):
Hi @AmiyaPiyush1, sorry for the delay in reviewing the PR.

I noticed that you force-pushed the recent commit. Could you please let me know the reason for doing that?

Also, once you make the changes, please make sure to address all reviewer comments so we can clearly see which points have been addressed.

A few tests are currently failing—could you please fix them (if it is related to your changes), update your branch with the latest upstream changes, and then reassign the PR back to me for review?

Please let me know if you have any questions.
Thank you!

AmiyaPiyush1 committed:
Merge branch 'develop' into fix-library-filter-23651

AmiyaPiyush1 commented:
Thank you @NITISH084 and @Nik-09 for the guidance and reviews! This was my first PR and I learned a lot about the merge process and resolving conflicts properly. Excited to contribute more!

AmiyaPiyush1 commented:
@NITISH084 

PTAL! at failed merge

AmiyaPiyush1 commented:
@NITISH084 

I have filed both the CI errors.

PTAL!

AmiyaPiyush1 commented:
E2E(navigation) error was covered in issue  - #23729

AmiyaPiyush1 commented:
@jayam04 Thanks for the heads up! From next time i will be more cautious.

AmiyaPiyush1 commented:
@NITISH084 

The merge queue failed due to two unrelated infrastructure flakes:

E2E (users): Failed with "Login page did not load" (timeout). I have verified the logs/screenshots and reported this in #24042.

Lighthouse perf (shard 1): Timed out after 45 minutes. which I think doesn't need an issue to be filed.

PTAL!

AmiyaPiyush1 commented:
@NITISH084 
PTAL!
